### PR TITLE
Chunk fetching APIs from Control Plane on startup

### PR DESCRIPTION
**Pull request by renuka-fernando**
### Purpose
Chunk retrieving deployments from controlplane to reduce the overhead to the controlplane.

### Automation tests
 - Unit tests added: Yes
 - Integration tests added: No

### Tested environments

Tested in Dev envirionment

---
#### Maintainers: Check before merge
- [x] Assigned 'Type' label
- [ ] Assigned the project
- [ ] Validated respective github issues
- [ ] Assigned milestone to the github issue(s)

#### With existing changes

```log
adapter-1   | 2024-02-28 06:20:01 INFO [main.go:39] - [main.startMicroGateway] [-] Starting Choreo Connect Adapter
adapter-1   | 2024-02-28 06:20:37 INFO [adapter.go:339] - [adapter.fetchAPIsOnStartUp] [-] Fetching APIs at startup with the received API UUID list (size: 3998)...
adapter-1   | 2024-02-28 06:27:28 INFO [adapter.go:350] - [adapter.fetchAPIsOnStartUp] [-] Fetching APIs at startup is completed...
```

#### With new changes
```log
adapter-1   | 2024-02-29 10:02:22 INFO [main.go:39] - [main.startMicroGateway] [-] Starting Choreo Connect Adapter
adapter-1   | 2024-02-29 10:02:57 INFO [adapter.go:343] - [adapter.fetchAPIsOnStartUp] [-] Fetching chunked APIs... [1/8]
adapter-1   | 2024-02-29 10:03:32 INFO [adapter.go:343] - [adapter.fetchAPIsOnStartUp] [-] Fetching chunked APIs... [2/8]
adapter-1   | 2024-02-29 10:04:07 INFO [adapter.go:343] - [adapter.fetchAPIsOnStartUp] [-] Fetching chunked APIs... [3/8]
adapter-1   | 2024-02-29 10:04:42 INFO [adapter.go:343] - [adapter.fetchAPIsOnStartUp] [-] Fetching chunked APIs... [4/8]
adapter-1   | 2024-02-29 10:05:20 INFO [adapter.go:343] - [adapter.fetchAPIsOnStartUp] [-] Fetching chunked APIs... [5/8]
adapter-1   | 2024-02-29 10:05:57 INFO [adapter.go:343] - [adapter.fetchAPIsOnStartUp] [-] Fetching chunked APIs... [6/8]
adapter-1   | 2024-02-29 10:06:32 INFO [adapter.go:343] - [adapter.fetchAPIsOnStartUp] [-] Fetching chunked APIs... [7/8]
adapter-1   | 2024-02-29 10:07:13 INFO [adapter.go:343] - [adapter.fetchAPIsOnStartUp] [-] Fetching chunked APIs... [8/8]
adapter-1   | 2024-02-29 10:08:50 INFO [adapter.go:351] - [adapter.fetchAPIsOnStartUp] [-] Fetching APIs at startup is completed...
```

### CODE REVIEW

File: adapter/internal/adapter/adapter.go [hunk 1]
```diff
@@ -328,9 +329,30 @@ OUTER:
...
+	if apiUUIDList == nil {
+		logger.LoggerMgw.Info("Fetching APIs at startup...")
+		fetchChunkedAPIsOnStartUp(conf, nil)
+	} else {
+		logger.LoggerMgw.Infof("Fetching APIs at startup with the received API UUID list (size: %d)...", len(apiUUIDList))
+
+		chunkedAPIUuidsList := utils.ChunkSlice(apiUUIDList, conf.ControlPlane.InitialFetch.ChunkSize)
+		for i, chunkedAPIUuids := range chunkedAPIUuidsList {
+			logger.LoggerMgw.Infof("Fetching chunked APIs... [%d/%d]", i+1, len(chunkedAPIUuidsList))
+			fetchChunkedAPIsOnStartUp(conf, chunkedAPIUuids)
```
[the remark this answers is not in the document]
Comment (renuka-fernando) [reply adapter/internal/adapter/adapter.go:344]: Yes, it retried only for the failed chunk

**Review by VajiraPrabuddhaka — APPROVED** [review]
LGTM

**Review by VirajSalaka — APPROVED** [review]

**Review by VirajSalaka — DISMISSED** [review]

**Review by VirajSalaka — APPROVED** [review]

**Review by malinthaprasan — APPROVED** [review]

**Comment by choreo-cicd** [comment]
[succeeded] Controlplane cluster : dev-deployment-v2 : 20240301.13

**Comment by choreo-cicd** [comment]
[succeeded] Dataplane(EastUS) cluster : dev-deployment-v2 : 20240301.13

**Comment by choreo-cicd** [comment]
[succeeded] Dataplane(NorthEU) cluster : dev-deployment-v2 : 20240301.13

**Comment by choreo-cicd** [comment]
[succeeded]  : dev-deployment-v2 : 20240301.13

**Comment by choreo-cicd** [comment]
[failed] Dataplane(NorthEU) cluster : stage-deployment-v2 : 20240304.2

**Comment by choreo-cicd** [comment]
[failed] Controlplane cluster : stage-deployment-v2 : 20240304.2

**Comment by choreo-cicd** [comment]
[succeeded]  : stage-deployment-v2 : 20240304.2

**Comment by choreo-cicd** [comment]
[failed] Dataplane(EastUS) cluster : stage-deployment-v2 : 20240304.2

**Comment by choreo-cicd** [comment]
[canceled] Dataplane(EastUS) cluster : stage-deployment-v2 : 20240304.2

**Comment by choreo-cicd** [comment]
[succeeded] Dataplane(NorthEU) cluster : stage-deployment-v2 : 20240304.2

**Comment by choreo-cicd** [comment]
[succeeded] Controlplane cluster : stage-deployment-v2 : 20240304.2

**Comment by choreo-cicd** [comment]
[succeeded]  : stage-deployment-v2 : 20240304.2

**Comment by choreo-cicd** [comment]
[] Dataplane(EastUS) cluster : stage-deployment-v2 : 20240304.2

**Comment by choreo-cicd** [comment]
[] Dataplane(NorthEU) cluster : stage-deployment-v2 : 20240304.2

**Comment by choreo-cicd** [comment]
[] Controlplane cluster : stage-deployment-v2 : 20240304.2

**Comment by choreo-cicd** [comment]
[]  : stage-deployment-v2 : 20240304.2

**Comment by choreo-cicd** [comment]
[succeeded] Dataplane(NorthEU) cluster : prod-deployment-v2 : 20240305.4

**Comment by choreo-cicd** [comment]
[succeeded]  : prod-deployment-v2 : 20240305.4

**Comment by choreo-cicd** [comment]
[succeeded] Controlplane cluster : prod-deployment-v2 : 20240305.4

**Comment by choreo-cicd** [comment]
[succeeded] Dataplane(EastUS) cluster : prod-deployment-v2 : 20240305.4

**Pull request merged**